csrc/scheduler/matmul_utils.cpp

```diff
@@ -186,6 +206,171 @@ inline bool initCoreHeuristics(
   return true;
 }
 
+bool fillDefaultHopperHeuristic(
+    MatmulParams* mparams,
+    const ProblemShape& problem_shape,
+    const mma_utils::TensorRolesMap& tensor_roles,
+    const size_t num_problems) {
+  const auto device_prop = at::cuda::getCurrentDeviceProperties();
+
+  const GemmTile instruction_tile = getMmaOpShape(mparams->mma_macro);
+  GemmTile warp_tile = {-1, -1, -1};
+  GemmTile cta_tile = {-1, -1, -1};
+
+  using DimType = decltype(GemmTile::m);
+
+  // We typically use larger macros on Hopper. By default we will set the
+  // warp tile equal to the macro and increase the CTA tile until we hit
+  // a limit. The limits are given by the maximum number of threads per CTA.
+
+  // TODO: it might be advantageous in some cases to issue multiple wgmma
+  // instructions per warp group
+  warp_tile = instruction_tile;
+
+  // The MmaOp output is a 32-bit float which requires one register per value
+
+  // The Hopper register file is 256KiB. We reduce this by a factor of 1/2 to
+  // account for overhead, since not all of the registers will hold MMA
+  // outputs.
+  const size_t max_registers_per_sm = device_prop->regsPerMultiprocessor / 2L;
```

Comment:
I wonder if `getRegPerThreadGivenThreadsPerSM` is more accurate than `registers_per_sm / 2`.
```suggestion
  // tma warp group + 2 * compute warp groups
  constexpr int64_t threads_per_sm = 384;
  const size_t max_registers_per_sm = getRegPerThreadGivenThreadsPerSM(threads_per_sm) * threads_per_sm;
```

Reply:
The number of threads per sm is determined by the number of warp groups, so I moved this down a bit.